pymemcache/ring.py

```diff
@@ -0,0 +1,173 @@
+# -*- coding: utf-8 -*-
+import sys
+import six
+
+"""
+    hash_ring
+    ~~~~~~~~~~~~~~
+    Implements consistent hashing that can be used when
+    the number of server nodes can increase or decrease (like in memcached).
+
+    Consistent hashing is a scheme that provides a hash table functionality
+    in a way that the adding or removing of one slot
+    does not significantly change the mapping of keys to slots.
+
+    More information about consistent hashing can be read in these articles:
+
+        "Web Caching with Consistent Hashing":
+            http://www8.org/w8-papers/2a-webserver/caching/paper2.html
+
+        "Consistent hashing and random trees:
+        Distributed caching protocols for relieving hot spots on the World Wide Web (1997)":
+            http://citeseerx.ist.psu.edu/legacymapper?did=38148
+
+
+    Example of usage::
+
+        memcache_servers = ['192.168.0.246:11212',
+                            '192.168.0.247:11212',
+                            '192.168.0.249:11212']
+
+        ring = HashRing(memcache_servers)
+        server = ring.get_node('my_key')
+
+    :copyright: 2008 by Amir Salihefendic.
+    :license: BSD
+"""
+
+import math
+import sys
+from bisect import bisect
+
+if sys.version_info >= (2, 5):
+    import hashlib
+    md5_constructor = hashlib.md5
+else:
+    import md5
+    md5_constructor = md5.new
+
+class HashRing(object):
```

Comment:
Did you write this code? If not, where did you get it? I just need to double check the licensing.